ref(otel): Add `ClientOptions.Instrumenter`

costela opened the pull request:
<!--

Hey, thanks for your contribution!

The Sentry team has finite resources and priorities that are not always visible on GitHub.
Please help us save time when reviewing your PR by following this two-step guide:

1. Is your PR a simple typo fix? __Click that green "Create pull request" button__!
2. For more complex PRs, please read https://github.com/getsentry/sentry-go/blob/master/CONTRIBUTING.md

-->

This small PR is a first attempt to fix the perceived issue in #678.

It extends `sentry.ClientOptions` with the `Instrumenter` parameter, analogous to other language SDKs. This can be optionally set to "otel" or "sentry" (the default).  

This new option is then respected by middlewares to optionally skip creating a sentry transaction, allowing the consumer to only use "pure" otel for traces.

### Considered alternatives:

Another option would be making the transaction created by the otel [span processor](https://pkg.go.dev/github.com/getsentry/sentry-go/otel#NewSentrySpanProcessor) visible to the middlewares. However, since the span processor cannot easily modify the request context, it becomes necessary to add additional helpers (e.g. a new "adapter" middleware) with the single task of injecting the otel-created transaction into the request context.


Fixes: #678 #788

costela commented:
The windows test failure seems unrelated? Maybe flaky?

As for the coverage decrease, I could add some tests for this case if the approach seems reasonable.

costela commented:
Sorry for the delay; vacation got in the way :wink: 

My javascript is pretty rusty (no :crab: pun intended), but AFAICT the js integration works slightly differently. It [always uses an existing span context](https://github.com/getsentry/sentry-javascript/blob/dc653d0adb95198daace6975b2343fd1d4443847/packages/opentelemetry-node/src/spanprocessor.ts#L34) (similarly to the proposed change in this PR), but it also has an extra [`instrumenter` option](https://github.com/getsentry/sentry-javascript/blob/dc653d0adb95198daace6975b2343fd1d4443847/packages/node/src/sdk.ts#L159), unlike the go integration (and this PR, for that matter).

The extra option seems a bit error-prone, compared to the sample-func approach in this PR: the code has to remember to check against known values, instead of explicitly passing exactly what we want it to do. But that's just my subjective (and superficial!) take on it.

costela commented:
Ok, third time is a charm!

The `sentry.SpanOption` doesn't quite work, because it can't modify it's received context. So we have to go with yet another middleware, which sits as an adapter between otel and `sentryhttp.Handler`. 
It sets the transaction created via the span processor in the context passed to `sentryhttp.Handler`, so that it detects its existence and can react accordingly.

PTAL

costela commented:
@cleptric can you please allow the tests to run? :pray:

costela commented:
FYI: :repeat: rebased on master, please unlock tests :pray:

cleptric commented:
Sorry that it took us a while to get back to you.

When using Otel, you shouldn't use any Sentry performance-related features. This is why we added the `instrumenter` option on other SDKs, which would disable all Sentry instrumentation. In Go, we opted not to add this option, given the manual nature of performance tracing.

So instead of using Sentry's `net/http` instrumentation, you would use https://github.com/open-telemetry/opentelemetry-go-contrib.

Our propagator considers incoming trace headers, so distributed tracing also works.

costela commented:
> So instead of using Sentry's net/http instrumentation, you would use https://github.com/open-telemetry/opentelemetry-go-contrib.

So you mean skipping the [http middleware](https://pkg.go.dev/github.com/getsentry/sentry-go/http)? But then we'd also throw out error/panic tracking?

If that's the intended way to work with OTEL, then maybe I should change this PR to instead add an option to `sentryhttp.Options` to skip traces?

cleptric commented:
This is a good point about panic tracking I haven't thought about.

IMO, we could wrap this code https://github.com/getsentry/sentry-go/blob/master/http/sentryhttp.go#L97-L112 into an if condition that checks for `ClientOptions.EnableTracing`, so no transactions are created.

This might make it a bit easier than adding a new integration option.

costela commented:
@cleptric you mean disabling `ClientOptions.EnableTracing`? I think that alone won't quite work either: then *all* transactions get dropped (even if the OTEL propagator decides to sample them).

We could maybe refactor the sampling logic to respect sample decisions even if `EnableTracing` is `false`, but I fear that might be too surprising to users.

In the end we want `sentryhttp.Handler` to only start a transaction if none has been started already (e.g. by the otel propagator), but unfortunately I haven't found a nice way to pass that informatino along. The propagator can't modify the request context and their otherwise unaware of each other. That's why I ended up with the current suggestion in this PR. :confused:

greywolve commented:
I think adding the `instrumenter` option also seems to be consistent with the other sdks.

costela committed:
feat: add TracesInstrumenter option

This enables instrumentation code (sentryhttp, sentrygin, etc) to skip
trace instrumentation while still performing panic recovery.

costela commented:
@cleptric @greywolve now this PR has an attempted implementation of the new `TracesInstrumenter` option. PTAL :pray:

cleptric commented on client.go:
```diff
@@ -133,6 +133,9 @@ type ClientOptions struct {
 	TracesSampler TracesSampler
+	// Which instrumentation to use for tracing. Either "sentry" or "otel" are supported.
+	// Setting this to "otel" will ignore TracesSampleRate and TracesSampler and assume sampling is performed by otel.
+	TracesInstrumenter string
```
We should call it `Instrumenter` as in other SDKs.

costela replied:
Sure thing! Done!

cleptric commented on client.go:
```diff
@@ -301,6 +304,19 @@ func NewClient(options ClientOptions) (*Client, error) {
 		options.MaxSpans = defaultMaxSpans
 	}
 
+	switch options.TracesInstrumenter {
+	case "":
+		options.TracesInstrumenter = "sentry"
+	case "sentry":
+		// noop
+	case "otel":
+		// sampling is performed by the OpenTelemetry SDK
+		options.TracesSampleRate = 1.0
+		options.TracesSampler = nil
+	default:
+		return nil, fmt.Errorf("invalid value for TracesInstrumenter (supported are 'sentry' and 'otel'): %q", options.TracesInstrumenter)
+	}
+
```
I'm not sure if we should do this. I would rather have people define this explicitly in their options.

costela replied:
you mean the `TracesSampleRate` and `TracesSampler`? Should we then error if the values are set to anything else? Doing neither seems a bit error-prone: `Instrumenter: "otel"` + `TracesSampleRate: 0` (default) would ignore otel sampling decisions and never send traces.

cleptric replied:
Are you sure it does this? The span processor actually takes the Otel sampling decision into account, taking priority over the traces sample rate and traces sampler.

costela replied:
I believe so. That might be a slightly different issue, though. One I had addressed in a [previous commit](https://github.com/getsentry/sentry-go/commit/9b18d57b992a06b384986d5e6faffef1871a293c#diff-27f9ebae19f3d3f9ca89ecaccdc8d8472ba81f86392efcb74e386532d26ca18a).
The current span processor considers only the sampled decision of the _parent_, so it only actually respects it if the span is part of an incoming trace. If it's a new trace, without a parent, the otel span might be sampled, but the sentry span remains at `SampleUndefined`, which then means we fall back to the client-wide sample rate.

[The current test](https://github.com/getsentry/sentry-go/blob/4c8f48a520c411e05569b35b0ade96310cdfb7af/otel/span_processor_test.go#L121) doesn't catch it because both the sentry client and the otel stack are initialized with "always trace", so they both arrive at the same decision, even though the processor isn't doing anything.

I just added a commit with the fixed test _and_ the fixed code. PTAL.

costela replied:
friendly ping :innocent:

cleptric replied:
I'll take a look later this week.

We made similar changes for node.js in https://github.com/getsentry/sentry-javascript/pull/9203, but I need to dive into this whole topic a bit deeper to give you actionable feedback.

costela replied:
friendly ping :innocent: (I know these things get lower prio; just wanna avoid it getting forgotten)

cleptric replied:
Yeah, sorry for that! I had to get some stuff shipped over on the PHP side of things, but the next few weeks will be again Go focus time for me.

cleptric reviewed (commented):
Added some comments. This also needs tests.

costela committed:
test: cover new instrumenter client option

costela committed:
fix(otel): also respect sampling on parentless spans

costela commented:
I want to add another perspective to this MR, that may mean the `ClientOptions.Instrumenter` is not enough: some client SDKs (e.g. sentry-cocoa) do not seem to support opentelemetry at all. These  transactions would be lost when speaking to a sentry+otel service. The service only reads otel headers, the client only sends sentry headers.

The suggestion with the additional middleware would not have this problem: incoming requests could use either sentry or otel propagation, the service code can leverage existing otel integrations (e.g. for [pgx](https://github.com/exaring/otelpgx/); shameless plug :see_no_evil:)

Unless sentry plans to support otel propagation on client sdks, it may be necessary to make the service side accept both. Or am I overlooking something?

cleptric commented:
We actually extract the Sentry trace header, see

https://github.com/getsentry/sentry-go/blob/master/otel/propagator.go#L77-L95

costela commented:
@cleptric ah, you're right of course. Appologies for the noise; disregard the last comment.

greywolve commented on client.go:
```diff
@@ -301,6 +304,14 @@ func NewClient(options ClientOptions) (*Client, error) {
+		options.Instrumenter = "sentry"
+	case "sentry", "otel": // noop
+	default:
+		return nil, fmt.Errorf("invalid value for TracesInstrumenter (supported are 'sentry' and 'otel'): %q", options.Instrumenter)
```
```suggestion
		return nil, fmt.Errorf("invalid value for Instrumenter (supported are 'sentry' and 'otel'): %q", options.Instrumenter)
```

costela replied:
oh, good catch! Thanks! :pray:

costela committed:
fix: report correct option in error

Co-authored-by: Oliver Powell <oliverpowell84@gmail.com>

greywolve commented:
@costela, thinking more about the otel changes specifically. I think you're right and your fix seems correct, but wouldn't that be better as a separate PR since it's addressing an additional concern?

greywolve commented:
LGTM, but I'll let @cleptric give the final thumbs up :)

costela commented:
hi!
just a little bi-monthly rebase + reminder about this PR :innocent:

costela commented:
maybe I can get this PR as a christmas present? :innocent: :santa: :christmas_tree: 

(just rebased :repeat:)

costela commented:
Rebased again.

Any chance a paying customer can get some feedback? Even if this isn't the right solution?

cleptric commented:
It's still on our radar!

rodolfoBee commented:
@cleptric [customer has requested](https://sentry.zendesk.com/agent/tickets/123251) an update on this PR. Is there any?

aldy505 commented:
@cleptric any updates for this?

giortzisg commented:
Closing this, since the span processor is [removed](https://github.com/getsentry/sentry-go/pull/1307) from 0.47.0 and on.
